File: eel-core/src/main/scala/io/eels/FilePattern.scala
package io.eels

import com.sksamuel.exts.Logging
import io.eels.util.HdfsIterator
import org.apache.hadoop.fs.{FileSystem, Path}

import scala.language.implicitConversions

object FilePattern {
  def apply(path: Path)(implicit fs: FileSystem): FilePattern = apply(path.toString())
  def apply(path: java.nio.file.Path)(implicit fs: FileSystem): FilePattern = apply(path.toAbsolutePath().toString(), { _ => true })

  implicit def stringToFilePattern(str: String)(implicit fs: FileSystem): FilePattern = FilePattern(str)
}

case class FilePattern(pattern: String,
                       filter: org.apache.hadoop.fs.Path => Boolean = { _ => true }) extends Logging {

  def isRegex(): Boolean = pattern.contains("*")
  def isDirectory(): Boolean = pattern.endsWith("/")

  def toPaths()(implicit fs: FileSystem): List[Path] = {
    val paths = if (isRegex) {

      val regex = new Path(pattern).getName.replace("*", ".*?")
      val dir = new Path(pattern).getParent
      logger.debug(s"File expansion will check path $dir for files matching $regex")

      HdfsIterator.remote(fs.listFiles(dir, false)).toList
        .map(_.getPath)
        .filter { path => path.getName.matches(regex) }
        .filter(filter)

    } else if (fs.isDirectory(new Path(pattern))) {

      val path = new Path(pattern.stripSuffix("/"))
      logger.debug(s"File expansion will search directory $path")
      HdfsIterator.remote(fs.listFiles(path, false)).map(_.getPath).toList.filter(fs.isFile).filter(filter)

    } else {
      List(new Path(pattern))
    }

    logger.debug(s"toPaths has returned ${paths.size} paths, first 5: ${paths.take(5).mkString(",")}")
    paths
  }

  def withFilter(p: Path => Boolean): FilePattern = copy(filter = p)
}

File: eel-core/src/main/kotlin/io/eels/FilePattern.kt
package io.eels

import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.Path
import java.io.File

data class FilePattern(val pattern: String,
                       val fs: FileSystem,
                       val filter: (Path) -> Boolean = { it -> true }) : Logging {

  constructor(path: Path, fs: FileSystem) : this(path.toString(), fs, { it -> true })

  constructor(path: java.nio.file.Path, fs: FileSystem) : this(path.toAbsolutePath().toString(), fs, { it -> true })

  val FileExpansionRegex = "(file:|hdfs:)?(?://)?(.*?)/\\*"

  fun isDirectory(): Boolean = pattern.endsWith("/*") || pattern.endsWith("/")
  fun isWildcard(): Boolean = pattern.endsWith("*")

  fun toPaths(): List<Path> = when {
    isDirectory() -> {
      val path = Path(pattern.removeSuffix("/*"))
        logger.debug("File expansion will check path: " + path)
      HdfsIterator(fs.listFiles(path, true)).map { it.path }.asSequence().toList().filter { fs.isFile(it) }.filter(filter)
    }
    isWildcard() -> {
      val path = Path(pattern.removeSuffix("*")).parent
      val regex = File(pattern).name.replace("*", ".*").toRegex()
        logger.debug("File expansion will check path: " + path)
      HdfsIterator(fs.listFiles(path, false)).asSequence().toList()
          .map { it.path }
          .filter { it.name.matches(regex) }
          .filter(filter)
    }
    else -> listOf(Path(pattern))
  }

  fun withFilter(p: (Path) -> Boolean): FilePattern = copy(filter = p)
}